### PR TITLE
MDEV-26015: Set DH param automatically

### PR DESCRIPTION
**Pull request by 9EOR9**
So far  MariaDB Server creates a DH (Diffie Hellman) parameter with a
fixed length (= 2048). This leads to the limitation that, for example, the
use of a DHE cipher suite requires a server certificate with an RSA key
of the same length.

To remedy this, the DH parameter is now automatically set by OpenSSL or WolfSSL
during the TLS handshake (taking into account the key length used
of the server certificate).

While WoldSSL supports this by default, OpenSSL has to activate this using the
SSL_CTX_set_dh_auto macro. OpenSSL versions < 1.0.2  doesn't support this feature
and use the previous implementation with a DH parameter with a fixed size.

Since command line client < 10.4.6 doesn't support tls_version parameter, testing
with DHE cipher suite was implemented in mysql_client_test.

- [x] *The Jira issue number for this PR is: MDEV-_____*

## Description
TODO: fill description here

## How can this PR be tested?
TODO: fill steps to reproduce here, if applicable,
      or remove the section

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

## Backward compatibility
TODO: fill details here, if applicable, or remove the section

### CODE REVIEW

**Review by vaintroub — CHANGES_REQUESTED** [review]
Could you make it compilable,  and pass tests on all platforms (including the Appveyor tests), then squash your commits, so it can be reviewed? BTW, There is no WolfSSL in the version which you decided to have as base for your commit. WolfSSL debuted in 10.4

**Comment by vaintroub** [comment]
image
that needs to be all green

File: vio/viosslfactories.c [hunk 1]
```diff
@@ -26,7 +26,12 @@
 static my_bool     ssl_algorithms_added    = FALSE;
 static my_bool     ssl_error_strings_loaded= FALSE;
 
-/* the function below was generated with "openssl dhparam -2 -C 2048" */
+#if !defined(WOLFSSL) && !defined(SSL_CTRL_SET_DH_AUTO)
```
Comment (dlenski) [inline vio/viosslfactories.c:29]: It appears that `#define SSL_CTRL_SET_DH_AUTO` was added considerably before OpenSSL v1.0.2: https://github.com/openssl/openssl/commit/09599b52d4e295c380512ba39958a11994d63401 ("Auto DH support."), first tagged in https://github.com/openssl/openssl/releases/tag/OpenSSL_1_1_0-pre1.

Will this feature _actually work_ with OpenSSL >=1.1.0-pre1 < 1.0.2?

Should there additionally be a check for `OPENSSL_VERSION_NUMBER < 0x10200000L`?

File: vio/viosslfactories.c [hunk 2]
```diff
@@ -269,25 +274,40 @@ new_VioSSLFd(const char *key_file, const char *cert_file,
...
+     WolfSSL chooses right DH public key automatically depending
+     on RSA key size of server certificate during TLS handshake.
+     For OpenSSL >=  1.0.2 this feature needs to be explicitly turned on.
+     OpenSSL versions < 1.0.2 will still use the old get_dh2048() method
+     to obtain and set DH public key (see also MDEV-26015).
+  */
   if (!is_client_method)
   {
-    dh=get_dh2048();
+#if !defined(SSL_CTRL_SET_DH_AUTO)
```
Comment (dlenski) [inline vio/viosslfactories.c:288]: Same question [as above](https://github.com/MariaDB/server/pull/1869/files#r721693775) regarding whether or not an explicit check for OpenSSL >=1.0.2 is needed.

**Comment by vaintroub** [comment]
@9EOR9 , this needs to pass on CI, before it can be reviewed. apparently, something is weird about Windows, as you can see from Appveyor output

**Review by vaintroub — CHANGES_REQUESTED** [review]
as commented earlier, this needs to be tested also on 10.4, and pass CI on Windows.
I think 10.4 as target should be fine, 10.2/3 are old and, we might not want to mess up with yassl compatibility.

**Comment by LinuxJedi** [comment]
Hi @9EOR9,

This appears to have been resolved elsewhere in 10.8.1 as part of [MDEV-25785](https://jira.mariadb.org/browse/MDEV-25785). Is it OK to close this PR now?

**Comment by LinuxJedi** [comment]
As my previous comment I will close this for now. Please feel free to reopen if you feel this is in error.

**Pull request closed without merging**